htsget-search/src/htsget/cram_search.rs

```diff
@@ -0,0 +1,324 @@
+    fn get_byte_ranges_for_header(
+        &self,
+        crai_index: &[crai::Record],
+    ) -> Result<BytesRange> {
+        // Assuming that the first index represents the first data container.
+        let first_record = crai_index.first().ok_or_else(|| HtsGetError::not_found("No entries in CRAI"))?;
+        Ok(BytesRange::default()
+            .with_start(Self::FILE_DEFINITION_LENGTH)
+            .with_end(first_record.offset())
+        )
+    }
+
+    /// Read header using storage options.
+    fn read_header(
+        &self,
+        key: &str,
+        header_bytes: BytesRange,
+    ) -> Result<(Reader<BufReader<File>>, Header)> {
+        let get_options = GetOptions::default().with_range(header_bytes);
+        let cram_path = self.storage.get(key, get_options)?;
+
+        let mut reader = File::open(cram_path)
+            .map(BufReader::new)
+            .map(noodles_cram::Reader::new)
+            .map_err(|_| HtsGetError::io_error("Reading CRAM"))?;
+
+        reader.read_file_definition().map_err(|_| HtsGetError::io_error("Reading CRAM file definition"))?;
+
+        let header = reader
+            .read_file_header()
+            .map_err(|_| HtsGetError::io_error("Reading CRAM header"))?
+            .parse()
+            .map_err(|_| HtsGetError::io_error("Parsing CRAM header"))?;
+
+        Ok((reader, header))
+    }
+
+    /// Get key for storage object.
+    fn get_keys_from_id(&self, id: &str) -> (String, String) {
+        let cram_key = format!("{}.cram", id);
+        let crai_key = format!("{}.crai", cram_key);
+        (cram_key, crai_key)
+    }
+
+    /// Returns mapped and placed unmapped ranges
+    fn get_byte_ranges_for_all_reads(
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |_| true,
+        )
+    }
+
+    /// Returns only unplaced unmapped ranges
+    fn get_byte_ranges_for_unmapped_reads(
+        &self,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+    ) -> Result<Vec<BytesRange>> {
+        Self::get_bytes_ranges(
+            crai_index,
+            cram_reader,
+            |record| {
+                record.reference_sequence_id().map_or_else(
+                    || true,
+                    |ref_seq| i32::from(ref_seq) == reference_sequence_id::UNMAPPED,
+                )
+            },
+        )
+    }
+
+    /// Returns reads for a given reference name and an optional sequence range
+    fn get_byte_ranges_for_reference_name(
+        &self,
+        reference_name: &str,
+        crai_index: &[crai::Record],
+        cram_reader: &mut Reader<BufReader<File>>,
+        cram_header: Header,
+    ) -> Result<Vec<BytesRange>> {
+        let maybe_cram_ref_seq = cram_header.reference_sequences().get_full(reference_name);
```

Comment:
If you only need the raw reference sequence ID, prefer [`get_index_of`](https://docs.rs/indexmap/1.7.0/indexmap/map/struct.IndexMap.html#method.get_index_of).

Reply:
I ended up backtracking on this change, because I needed the reference sequence to simplify checking overlapping ranges in the query start/end. Similar to https://github.com/umccr/htsget-rs/blob/8ce397649da93c4624c617c0f415fd64c4b86951/htsget-search/src/htsget/bam_search.rs#L190.